### PR TITLE
Implement signing in FxOS Add-ons (bug 1198696)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1198696

### CODE REVIEW

File: mkt/extensions/migrations/0002_extension_add_uuid.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
```
Comment (chuckharmston) [inline mkt/extensions/migrations/0002_extension_add_uuid.py]: These are great work.

Comment (diox) [inline mkt/extensions/migrations/0002_extension_add_uuid.py]: Can't take the credit for it, the idea comes from the django docs. I was about to write a single python migration, the same kind that would sometimes break back when we were using schematic. Was pretty happy to read how to do it properly in the docs :)

**Comment by chuckharmston** [comment]
Not super comfortable with the signing plans for addons, so I'll defer the r+ to somebody who knows expectations. Looks great though: endpoints, migrations, tests. Well done.

**Comment by washort** [comment]
r+

File: mkt/extensions/migrations/0003_populate_extension_uuid.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+def add_uuid_to_exising_extensions(apps, schema_editor):
+    Extension = apps.get_model('extensions', 'Extension')
+    extensions = Extension.objects.all()
+    for extension in extensions:
+        extension.uuid = extension._meta.get_field('uuid')._create_uuid()
```
Comment (robhudson) [inline mkt/extensions/migrations/0003_populate_extension_uuid.py]: Super secret internal fields and methods. :scream:

File: mkt/extensions/migrations/0003_populate_extension_uuid.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('extensions', '0002_extension_add_uuid'),
+    ]
+
+    operations = [
+        # Update the uuid field on each row.
+        migrations.RunPython(
+            add_uuid_to_exising_extensions,
```
Comment (robhudson) [inline mkt/extensions/migrations/0003_populate_extension_uuid.py]: This is just an arg to the `RunPython` call right? It seems odd to include the trailing comma to me.

File: mkt/extensions/models.py [hunk 4]
```diff
@@ -140,18 +155,52 @@ def is_public(self):
...
+            # 'version' should be an integer and should be monotonically
+            # increasing.
+            'version': self.file_version
+        })
+        with statsd.timer('extensions.sign'):
+            try:
+                # This will read the file from self.file_path, generate a
+                # signature and write the signed file to self.signed_file_path.
+                sign_app(private_storage.open(self.file_path),
+                         self.signed_file_path, ids)
```
Comment (robhudson) [inline mkt/extensions/models.py]: `sign_app` will work for both packaged apps and extensions?

**Comment by robhudson** [comment]
nice work

File: mkt/extensions/migrations/0003_populate_extension_uuid.py (hunk 2 above)
Comment (diox) [inline mkt/extensions/migrations/0003_populate_extension_uuid.py]: Yeah. It's not that secret, really all it does is a `uuid.uuid4()` call. I could replace it - I wanted to use what the field really uses behind the scenes

File: mkt/extensions/models.py (hunk 4 above)
Comment (diox) [inline mkt/extensions/models.py]: Yeah, it doesn't care about the manifest, and they are both zips, signed the same way. It *should* work on the device, but I really have no idea at this point : I need to have the manifest and signing working to test it.

**Pull request merged**